### PR TITLE
Addded mig support for 3 CMake syntaxes

### PR DESCRIPTION
**Pull request by TejaX-Alaghari**
Added migration support for 3 CMake syntaxes:

1. find_program
2. find_library
3. link_directories

### CODE REVIEW

**Review by abhilash1910 — COMMENTED** [review]
LGTM! @tomflinda  @zhimingwang36  could you please help with review and merge process.

**Review by tomflinda — APPROVED** [review]
LGTM

**Review by zhiweij1 — APPROVED** [review]

File: clang/test/dpct/cmake_migration/case_061/expected.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+find_library(CUDA_LIB "" PATH ${prefix_path})
```
Comment (zhimingwang36) [inline clang/test/dpct/cmake_migration/case_061/expected.txt]: 2nd parameter of find_library() Specify one or more possible names for the library.   What expect to do if the name is empty?  should we remove it?

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml [hunk 2]
```diff
@@ -918,6 +942,225 @@
...
+  Out: find_library(${var_lib} ${libs})
+  Subrules:
+    libs:
+      In: nppc_static
+      Out: |
+        ""
+      MatchMode: Full
+      RuleId: "remove_find_library_nppc_static"
+
+- Rule: rule_find_library_nppicc_static
```
Comment (zhimingwang36) [inline clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml:1050]: miss test case for rule L1074?

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml [hunk 3]
```diff
@@ -918,6 +942,225 @@
...
+  Subrules:
+    libs:
+      MatchMode: Full
+      In: nvToolsExt
+      Out: |
+        ""
+      RuleId: "remove_find_library_nvToolsExt"
+
+- Rule: rule_find_library_cufftMp
+  Kind: CMakeRule
```
Comment (zhimingwang36) [inline clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml:978]: no test case for this rule?

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml [hunk 4]
```diff
@@ -918,6 +942,225 @@
...
+  Out: find_library(${var_lib} ${libs})
+  Subrules:
+    libs:
+      In: cufftMp
+      Out: |
+        ""
+      MatchMode: Full
+      RuleId: "remove_find_library_cufftMp"
+
+- Rule: rule_find_library_nvshmem
```
Comment (zhimingwang36) [inline clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml:991]: no test case for this rule?

File: clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml [hunk 5]
```diff
@@ -918,6 +942,225 @@
...
+      MatchMode: Full
+      RuleId: "remove_find_library_libnvjpeg_static"
+
+- Rule: rule_find_library_nvjpeg
+  Kind: CMakeRule
+  Priority: Fallback
+  MatchMode: Partial
+  CmakeSyntax: find_library_nvjpeg
+  In: find_library(${var_lib} ${libs})
+  Out: find_library(${var_lib} ${libs})
```
Comment (zhimingwang36) [inline clang/tools/dpct/DpctOptRules/cmake_script_migration_rule.yaml:1131]: please add test case for each new added rule.

File: clang/test/dpct/cmake_migration/case_063/input.cmake [hunk 6]
```diff
@@ -0,0 +1,9 @@
+find_program(CUDAToolkit_NVCC_EXECUTABLE nvcc ${CUDAToolkit_BIN_DIR})
+
+find_program(CUDAToolkit_NVCC_EXECUTABLE
+    NAMES nvcc nvcc.exe
```
Comment (zhimingwang36) [inline clang/test/dpct/cmake_migration/case_063/input.cmake:6]: what happen if nvcc_other, will it be replace by $(CMAKE_CXX_COMPILER)_other ?

Comment (zhimingwang36) [reply clang/test/dpct/cmake_migration/case_063/input.cmake:6]: I think partial match is not safe here.

Comment (abhilash1910) [reply clang/test/dpct/cmake_migration/case_063/input.cmake:6]: Yes , I think for nvcc_exe can be mapped to $(CMAKE_CXX_COMPILER) ?

File: clang/test/dpct/cmake_migration/case_061/expected.txt (hunk 1 above)
Comment (TejaX-Alaghari) [reply clang/test/dpct/cmake_migration/case_061/expected.txt]: If the lib name is empty, the var is set to **NOFOUND** value. No error is thrown.
Removing the entire statement would result in errors for the var not being defined.

File: clang/test/dpct/cmake_migration/case_063/input.cmake (hunk 6 above)
Comment (TejaX-Alaghari) [reply clang/test/dpct/cmake_migration/case_063/input.cmake:6]: Lib names of form **nvcc_other** will not be migrated to "${CMAKE_CXX_COMPILER}_other"

As the **MatchMode** is **Full**, the Input string is matched for the entire CMake arg until a valid separator (like space) is encountered

File: clang/test/dpct/cmake_migration/case_061/expected.txt (hunk 1 above)
Comment (zhimingwang36) [reply clang/test/dpct/cmake_migration/case_061/expected.txt]: ok

File: clang/test/dpct/cmake_migration/case_063/input.cmake (hunk 6 above)
Comment (zhimingwang36) [reply clang/test/dpct/cmake_migration/case_063/input.cmake:6]: if nvcc.other can be migrated to ${...}.other,  then  it is a bug, (partial match).

Comment (abhilash1910) [reply clang/test/dpct/cmake_migration/case_063/input.cmake:6]: Please re-review once.

**Review by zhimingwang36 — APPROVED** [review]

**Pull request merged**